### DISCUSSION TITLE
how to add our own update site url to the P2 Install/Update-> Trust -> Authorities?

### DISCUSSION THREAD

**Discussion by jcompagner** [eclipse-equinox/p2#302] [General]
I am trying to find out how i can add our own url there when people install our product and then want to update our product from that url 

our product does see the defaults now: 

AuthorityChecker:
```
	public Set<URI> getPreferenceTrustedAuthorities() {
		var result = new LinkedHashSet<URI>();
		var preferences = getEnngineProfilePreferences();
		if (preferences != null) {
			String defaultValue = EngineActivator.getContext().getProperty("p2.trustedAuthorities");
			if (defaultValue == null) {
				defaultValue = "https://download.eclipse.org https://archive.eclipse.org";
			} else {
				defaultValue = defaultValue.replace(',', ' ');
			}
			for (var value : preferences.get(TRUSTED_AUTHORITIES_PROPERTY, defaultValue).split("\\s+")) {
				try {
					if (!value.isBlank()) {
						result.add(new URI(value));
					}
				} catch (URISyntaxException e) {
					//$FALL-THROUGH$
				}
			}
		}
		return result;
	}
```

but i am wondering what i need to configure somewhere to also add our update site url

Because i guess, after that i guess unsigned jars are not reported like that? because the https url with a valid certificate will be seen as trusted (everything that is downloaded from it?)

it seems there is a "p2.trustedAuthorities" property but that is just an environment variable or something?

**Comment by merks** [comment]
Yes, it's effectively just an environment variable fetched in a proper OSGi way.  So in your *.product file you can specify the system property with a comma (or space) separated list of authorities like this
```
      <vmArgs>-Xms256M
-Xmx2048M
-Dp2.trustedAuthorities=https://my.org,https://download.eclipse.org
      </vmArgs>
```
This **only** affects the prompting of authorities, i.e., the source of content: 

https://eclipse.dev/eclipse/news/4.28/platform.php#Security

This is entirely orthogonal to prompting about signed content.  You should sign your content.

**Reply by jcompagner** [reply]
the problem with sign your content is that a lot is now more and more done straight from maven sites 
(so the target platform has mvn bundles) 

So yes we are signing all our code, but our application is a combination of our own plugins with eclipse shipped plugins and a lot of other 3rd party stuff

i had hoped that those trusted authorities could be something like: ok if it comes from this site (our p2 product update site) then just accept it even if they are unsigned,

Because the problem is signing all those 3rd party jars.. then i adjust the jar itself (make may own copy) but it has the same bundle name and version as potentially others... (orbit did this, but that is more and more slimmed down now, to not repackage mvn stuff that are correct osgi bundles)

**Reply by laeubi** [reply]
> then i adjust the jar itself (make may own copy) 

That's why PGP is the better way here, so there is no need to "make a modified copy"...

**Reply by jcompagner** [reply]
yes but i still don't know how to really set this up that it works simple on locally made builds or on stuff in the cloud..
Last time i checked i need to install all kind of native tools on the system, (but i do plan to look at it again)

If there is just a Tycho mojo where i can just give some config to and that then generated it all for me (just like jar signing) then i will try to move over quickly... But last time i checked it was not so simple to setup...

Will this also fix this problem? https://github.com/eclipse-equinox/p2/discussions/333 will the p2 repo then include all the right keys?

**Reply by laeubi** [reply]
I think @merks has recently simplified the PGP setup for platform at laest...

**Reply by merks** [reply]
Here's how we make sure that everything in the SimRel repository is PGP signed if it is not otherwise considered jar signed by an anchored certificate:

- https://github.com/eclipse-simrel/simrel.build/blob/28ffb677b6cf318afb83db071204058243174356/pom.xml#L100-L122
- https://github.com/eclipse-simrel/simrel.build/blob/28ffb677b6cf318afb83db071204058243174356/Jenkinsfile#L78-L92

**Reply by jcompagner** [reply]
thx but i if look at the mojo i see this:

https://tycho.eclipseprojects.io/doc/latest/tycho-gpg-plugin/sign-p2-artifacts-mojo.html#executable

thats all kind of native stuff right that needs to be installed?

**Reply by merks** [reply]
It's configured to use "bc" for signing so nothing native needs to be installed but rather it uses bouncycastle libraries directly from within the same process:

https://tycho.eclipseprojects.io/doc/latest/tycho-gpg-plugin/sign-p2-artifacts-mojo.html#signer

The documentation could be improved...

**Reply by jcompagner** [reply]
ah thx, that was not directly clear to me,

But still we need to make a private/public key that we use that is then registered at a key server?
(https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Fguide%2Fp2_pgp.html)

**Reply by merks** [reply]
Yes you definitely need to create the public/ private keys and for that you do need some tools.  There is no requirement to register it with a key server but users will perhaps find it more trustworthy if that's the case.  With the bc signer, you can "stuff" the private key and the password into system properties to be used in the Tycho build without special tools needing to be installed for the build server...
